File: python/quick.py
import time
import sys

def readNumbers(file, size):
    vet = []
    i = 0
    f = open(file, "r")
    for line in f:
        if i > size:
            break
        vet.append(int(line))
        i = i+1
    f.close()
    return vet

def quicksort(A, min, max):
    if min < max:
        q = partition(A, min, max)
        quicksort(A, min , q-1)
        quicksort(A, q+1 , max)

def partition(A, min, max):
    x = i = j = aux = 0
    x = A[max]
    i = min-1
    for j in range(min, max):
        if(A[j] <= x):
            i+=1
            #exchange(&A[i], &A[j]);
            aux = A[i]
            A[i] = A[j]
            A[j] = aux
    #exchange A[i+1], A[max]
    aux = A[i+1]
    A[i+1] = A[max]
    A[max] = aux
    return i+1

def orderNumbers(size):
    i = 0
    vet = []
    while i<size:
        vet.append(i)
        i+=1
    return vet

def reverseOrderNumbers(size):
    i = 0
    vet = []
    while size>0 :
        vet.append(size)
        size-=1
    return vet;

# main
REPETITION = 20
SIZE = 50000
vet = []
sys.setrecursionlimit(99999999)
print("QUICK SORT (Size=" + str(SIZE) + ", rept="+ str(REPETITION) + ")")
for x in range(REPETITION):
    #vet = readNumbers("random_numbers.txt", SIZE)
    vet = orderNumbers(SIZE)
    
    begin = int(round(time.time() * 1000))
    quicksort(vet, 0, len(vet)-1)
    end = int(round(time.time() * 1000))

    print (end-begin)

"""print ("Sorted array is:")
for i in range(len(vet)):
    print ("%d" %vet[i])
"""

File: c/quick.c
#include<stdio.h>
#include<stdlib.h>
#include<limits.h>
#include<time.h>

#define REPETITION 20
#define SIZE 10000

void exchange(int *a, int *b) {
    int aux = *a;
    *a = *b;
    *b = aux;
}

int partition(int A[], int min, int max) {
    int x, i, j;
    x = A[max];
    i = min-1;
    for(j=min; j<max; j++) {
        if(A[j] <= x) {
            i++;
            exchange(&A[i], &A[j]);
        }
    }
    exchange(&A[i+1], &A[max]);
    return i+1;
}

void quicksort(int A[], int min, int max) {
    if(min < max) {
        int q = partition(A, min, max);
        quicksort(A, min , q-1);
        quicksort(A, q+1 , max);
    }
}

int * readNumbers(char *file, int size) {
    int i=0, aux,*vet = malloc(sizeof(int) * size);
    FILE *f = fopen(file, "r");
    while(EOF!=fscanf(f, "%d", &aux) && i<size)
        vet[i++] = aux;
    fclose(f);
    return vet;
}

int * orderNumbers(int size) {
    int i=0, aux, *vet = malloc(sizeof(int) * size);
    while(i<size) {
        vet[i] = i++;
    }
    return vet;
}

int * reverseOrderNumbers(int size) {
    int i=0, aux=size, *vet = malloc(sizeof(int) * size);
    while(i<size){
        vet[i++] = aux--;
    }
    return vet;
}

void execute(int qtd, char order) {
    int i=0, *vet;
    double time_spent;
    clock_t end, begin;
    printf("QUICK SORT (Size=%d, rept=%d) Order=%c\n", qtd, REPETITION, order);
    for(i=0; i<REPETITION; i++) {
	if(order == 'R')
		vet = reverseOrderNumbers(qtd);
	else if(order == 'O')
		vet = orderNumbers(qtd);
	else if(order == 'F')
		vet = readNumbers("random_numbers.txt", qtd);
	else
		return;

        begin = clock();
        quicksort(vet, 0, qtd-1);
        end = clock();

        time_spent = ((double)(end - begin) / CLOCKS_PER_SEC) * 1000;
        printf("%.f\n", time_spent);
        free(vet);
     }
}
int main() {
//    execute(10000, 'R');
//    execute(100000, 'R');
//    execute(200000, 'R');
   // execute(1000000, 'O');
   // execute(200000, 'O');
   // execute(200000, 'F');
    execute(1000000, 'O');

/**
    int i=0, *vet;
    double time_spent;
    clock_t end, begin;
    printf("QUICK SORT (Size=%d, rept=%d)\n", SIZE, REPETITION);
    for(i=0; i<REPETITION; i++) {
        vet = readNumbers("random_numbers.txt", SIZE);

        begin = clock();
        quicksort(vet, 0, SIZE-1);
        end = clock();

        time_spent = ((double)(end - begin) / CLOCKS_PER_SEC) * 1000;
        printf("%.f\n", time_spent);
        free(vet);
     }
**/
    /**
    int B[100];
    for (j=0, k=100; j<100; j++){
        B[j] = k--;
    }
    printf("\n");
    quicksort(B, 0, 99);
    for (j=0, k=100; j<100; j++){
        printf("%d ", B[j]);
    }
    **/
    return 0;
}
